File: backend/src/main/java/com/cakeplatform/api/modules/product/Product.java
package com.cakeplatform.api.modules.product;

import com.cakeplatform.api.modules.shop.Shop;
import jakarta.persistence.*;
import lombok.Data;
import com.fasterxml.jackson.annotation.JsonIgnore;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "products")
@Data
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shop_id", nullable = false)
    private Shop shop;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private ProductCategory category;

    @Column(nullable = false)
    private String name;

    private String description;

    @Column(columnDefinition = "TEXT")
    private String ingredients;

    @Column(columnDefinition = "TEXT")
    private String allergens;

    @Column(nullable = false)
    private BigDecimal price;

    @Column(name = "original_price")
    private BigDecimal originalPrice;

    @Column(name = "allow_egg_choice", nullable = false)
    private Boolean allowEggChoice = false;

    @Column(name = "egg_preference_default")
    private String eggPreferenceDefault = "EGGLESS";

    @Column(name = "eggless_price_diff")
    private BigDecimal egglessPriceDiff = BigDecimal.ZERO;

    @Column(name = "image_url")
    private String imageUrl;

    @Column(nullable = false)
    private Boolean availability = true;

    @Column(nullable = false)
    private String status = "ACTIVE";

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("displayOrder ASC, createdAt ASC")
    private List<ProductImage> images = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("displayOrder ASC, createdAt ASC")
    private List<ProductHighlight> highlights = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("displayOrder ASC, id ASC")
    private List<ProductVariant> variants = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ProductAddon> addons = new ArrayList<>();

    @Transient
    @com.fasterxml.jackson.annotation.JsonProperty("categoryId")
    public Long getCategoryId() {
        return category != null ? category.getId() : null;
    }

    @Transient
    @com.fasterxml.jackson.annotation.JsonProperty("categoryName")
    public String getCategoryName() {
        return category != null ? category.getName() : null;
    }

    @Transient
    @com.fasterxml.jackson.annotation.JsonProperty("shopId")
    public Long getShopId() {
        return shop != null ? shop.getId() : null;
    }
}

File: frontend_v2/types/product.ts
export interface Category {
  id: number;
  shopId: number;
  name: string;
  slug?: string;
  displayOrder: number;
  productCount: number;
  createdAt?: string;
  updatedAt?: string;
}

export interface CategoryRequest {
  name: string;
  displayOrder?: number;
}

export interface ProductImage {
  id?: number;
  imageUrl: string;
  displayOrder: number;
  altText?: string;
}

export interface ProductHighlight {
  id?: number;
  highlightText: string;
  displayOrder: number;
}

export interface ProductVariant {
  id?: number;
  name: string;
  price: number;
  originalPrice?: number | null;
  imageUrl?: string | null;
  description?: string | null;
  displayOrder?: number;
  variantType?: string;
  isAvailable?: boolean;
}

export interface ProductAddon {
  id?: number;
  name: string;
  price: number;
  isAvailable?: boolean;
}

export interface Product {
  id: number;
  shopId: number;
  name: string;
  description: string;
  price: number;
  originalPrice?: number | null;
  categoryId?: number | null;
  categoryName?: string | null;
  category?: string;
  imageUrl?: string;
  isEggless: boolean;
  allowEggChoice?: boolean;
  eggPreferenceDefault?: 'EGGLESS' | 'REGULAR' | string;
  egglessPriceDiff?: number;
  inStock: boolean;
  availability?: boolean;
  status?: string;
  preparationTimeHours?: number;
  weightGrams?: number;
  ingredients?: string | null;
  allergens?: string | null;
  images?: ProductImage[];
  highlights?: ProductHighlight[];
  variants?: ProductVariant[];
  addons?: ProductAddon[];
  rating?: number;
  reviewCount?: number;
  totalReviews?: number;
}

export interface CreateProductRequest {
  name: string;
  description?: string;
  ingredients?: string | null;
  allergens?: string | null;
  price: number;
  originalPrice?: number | null;
  categoryId?: number | null;
  category?: string;
  imageUrl?: string;
  isEggless?: boolean;
  allowEggChoice?: boolean;
  eggPreferenceDefault?: string;
  egglessPriceDiff?: number;
  inStock?: boolean;
  availability?: boolean;
  preparationTimeHours?: number;
  weightGrams?: number;
  images?: ProductImage[];
  highlights?: ProductHighlight[];
  variants?: ProductVariant[];
  addons?: ProductAddon[];
}
